Add cloud functions to GCP plugin

ekinanp opened the pull request:
Resolves #578

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add cloud functions to GCP plugin

Resolves #578

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on plugin/gcp/cloudFunction.go:
```diff
@@ -0,0 +1,55 @@
+
+func newCloudFunction(function *cloudfunctions.CloudFunction, service cloudFunctionsProjectService) *cloudFunction {
+	cf := &cloudFunction{
+		EntryBase: plugin.NewEntry(function.Name),
```
Looks like cloud function names are of the form `project/<project_id>/location/<region>/functions/<function_name>`. Given this, how should we organize them? I'm thinking `<region>::<function_name>`.

MikaelSmith replied:
VM instances are also in a specific region/zone. As long as name uniqueness is enforced across zones I didn't put it in.

ekinanp commented on plugin/gcp/cloudFunction.go:
```diff
@@ -0,0 +1,55 @@
+
+func (cf *cloudFunction) Delete(ctx context.Context) (bool, error) {
+	_, err := cf.service.Projects.Locations.Functions.Delete(cf.Name()).Context(ctx).Do()
+	return false, err
```
Cloud functions have some shut down stuff that needs to happen before they're deleted, which is why this is false.

ekinanp committed:
Add pubsubDir schema to GCP project's child schemas

Signed-off-by: Enis Inan <enis.inan@puppet.com>

MikaelSmith commented on plugin/gcp/cloudFunction.go:
```diff
@@ -0,0 +1,55 @@
+		EntryBase: plugin.NewEntry(function.Name),
+		service:   service,
+	}
+	mtime, err := time.Parse(time.RFC3339, function.UpdateTime)
```
Huh, only some services add a creation timestamp so it's in "Metadata".

ekinanp replied:
Yeah, it's weird.

MikaelSmith commented on plugin/gcp/cloudFunctionLog.go:
```diff
@@ -0,0 +1,182 @@
+		if len(s.currentEntries) > 0 {
+			break
+		}
+		time.Sleep(2 * time.Second)
```
Why sleep for 2 seconds?

MikaelSmith replied:
Oh, because copy will busyread.

MikaelSmith replied:
I wish there were a better way, but I don't see it.

ekinanp committed:
Omit project and region from the cloud function name

We should never have to include the project. We'll include the region if
someone raises an issue.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Updated to omit project + region from function name.
